File: DSA/Linkedlist/Code/IntroductionToLinkedList.py
# Introduction to Linked List in Python
# Basic structure and node operations

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None
    
    def insert_at_beginning(self, data):
        """Insert at beginning O(1)"""
        new_node = Node(data)
        new_node.next = self.head
        self.head = new_node
    
    def insert_at_end(self, data):
        """Insert at end O(n)"""
        new_node = Node(data)
        
        if self.head is None:
            self.head = new_node
            return
        
        temp = self.head
        while temp.next is not None:
            temp = temp.next
        
        temp.next = new_node
    
    def display(self):
        """Display linked list"""
        elements = []
        temp = self.head
        while temp is not None:
            elements.append(str(temp.data))
            temp = temp.next
        print(" -> ".join(elements) + " -> NULL")
    
    def count_nodes(self):
        """Count total nodes"""
        count = 0
        temp = self.head
        while temp is not None:
            count += 1
            temp = temp.next
        return count
    
    def search(self, key):
        """Search for element"""
        temp = self.head
        while temp is not None:
            if temp.data == key:
                return True
            temp = temp.next
        return False
    
    def get_at_index(self, index):
        """Get element at index"""
        temp = self.head
        for _ in range(index):
            if temp is None:
                return None
            temp = temp.next
        return temp.data if temp else None

# Test
ll = LinkedList()

# Create linked list: 10 -> 20 -> 30 -> NULL
ll.insert_at_end(10)
ll.insert_at_end(20)
ll.insert_at_end(30)

print("Linked List: ", end="")
ll.display()

# Insert at beginning
ll.insert_at_beginning(5)
print("After inserting 5 at beginning: ", end="")
ll.display()

# Insert at end
ll.insert_at_end(40)
print("After inserting 40 at end: ", end="")
ll.display()

# Count nodes
print(f"Number of nodes: {ll.count_nodes()}")

# Search
print(f"Search 20: {'Found' if ll.search(20) else 'Not Found'}")
print(f"Search 50: {'Found' if ll.search(50) else 'Not Found'}")

# Get at index
print(f"Element at index 0: {ll.get_at_index(0)}")
print(f"Element at index 2: {ll.get_at_index(2)}")

print("\nComplexity Analysis:")
print("Create Node: O(1)")
print("Insert at beginning: O(1)")
print("Insert at end: O(n)")
print("Search: O(n)")
print("Display/Traverse: O(n)")
print("Space: O(n)")

File: DSA/Linkedlist/Code/IntroductionToLinkedList.c
#include <stdio.h>
#include <stdlib.h>

// Introduction to Linked List in C
// Basic structure and node operations

typedef struct Node {
    int data;
    struct Node* next;
} Node;

// Create a new node
Node* createNode(int data) {
    Node* node = (Node*)malloc(sizeof(Node));
    node->data = data;
    node->next = NULL;
    return node;
}

// Insert at beginning
Node* insertAtBeginning(Node* head, int data) {
    Node* newNode = createNode(data);
    newNode->next = head;
    return newNode;
}

// Insert at end
Node* insertAtEnd(Node* head, int data) {
    Node* newNode = createNode(data);
    
    if (head == NULL)
        return newNode;
    
    Node* temp = head;
    while (temp->next != NULL)
        temp = temp->next;
    
    temp->next = newNode;
    return head;
}

// Display linked list
void display(Node* head) {
    Node* temp = head;
    while (temp != NULL) {
        printf("%d -> ", temp->data);
        temp = temp->next;
    }
    printf("NULL\n");
}

// Count nodes
int countNodes(Node* head) {
    int count = 0;
    Node* temp = head;
    while (temp != NULL) {
        count++;
        temp = temp->next;
    }
    return count;
}

// Search element
int search(Node* head, int key) {
    Node* temp = head;
    while (temp != NULL) {
        if (temp->data == key)
            return 1;
        temp = temp->next;
    }
    return 0;
}

// Free memory
void freeList(Node* head) {
    Node* temp;
    while (head != NULL) {
        temp = head;
        head = head->next;
        free(temp);
    }
}

int main() {
    Node* head = NULL;
    
    // Create linked list: 10 -> 20 -> 30 -> NULL
    head = insertAtEnd(head, 10);
    head = insertAtEnd(head, 20);
    head = insertAtEnd(head, 30);
    
    printf("Linked List: ");
    display(head);
    
    // Insert at beginning
    head = insertAtBeginning(head, 5);
    printf("After inserting 5 at beginning: ");
    display(head);
    
    // Insert at end
    head = insertAtEnd(head, 40);
    printf("After inserting 40 at end: ");
    display(head);
    
    // Count nodes
    printf("Number of nodes: %d\n", countNodes(head));
    
    // Search
    printf("Search 20: %s\n", search(head, 20) ? "Found" : "Not Found");
    printf("Search 50: %s\n", search(head, 50) ? "Found" : "Not Found");
    
    printf("\nComplexity Analysis:\n");
    printf("Create Node: O(1)\n");
    printf("Insert at beginning: O(1)\n");
    printf("Insert at end: O(n)\n");
    printf("Search: O(n)\n");
    printf("Display/Traverse: O(n)\n");
    printf("Space: O(n)\n");
    
    freeList(head);
    return 0;
}
